### ISSUE TITLE
erlcloud_ddb2:global_table_status/0 doesn't exist?

### ISSUE THREAD

**Issue by paulo-ferraz-oliveira** [erlcloud/erlcloud#483]
dialyzer just complained about this in 3.1.1: erlcloud_ddb2:global_table_status/0

**Comment by nalundgaard** [comment]
Looks like my mistake, I namespaced a type in the erlcloud_ddb2.hrl file. I will PR a fix.

**Comment by paulo-ferraz-oliveira** [comment]
Thanks.

We've also detected that e.g. erlcloud_ec2.hrl uses type datetime() but doesn't include erlcloud.hrl (where the type is defined).

Some other hrl files do (include erlcloud.hrl), like erlcloud_as.hrl, but there's probably a better approach to avoid cyclic dependencies.

What if we (this breaks interface!) declare and export types only from .erl files. We can then have .hrl files as stand-alone.

So, we could have datetime in erlcloud_aws.erl or erlcloud.erl and then export it from there and import it (fully qualified) as erlcloud:datetime() where required.

**Comment by paulo-ferraz-oliveira** [comment]
Another way some other projects have found to go around this issue is to declare a _types.erl module, e.g. erlcloud_types.erl where all types would be defined.

**Comment by paulo-ferraz-oliveira** [comment]
datetime() is used in:
* erlcloud_as.hrl
* erlcloud_ec2.hrl
* erlcloud_mon.hrl
* erlcloud_mturk.hrl

and it's actually the only used content from erlcloud.hrl, since it also defines -type proplist() :: proplists:proplist(). but this is unused (at least internally).

**Comment by motobob** [comment]
please feel free to send a PR to fix datetime().
including that hlr should be of cyclic harm. 

in general, @paulo-ferraz-oliveira do i understand correct that your top level project requires dializer passing?

**Comment by nalundgaard** [comment]
@paulo-ferraz-oliveira in general I think I would agree with exporting shared types from modules correspondent to the headers they are in and referencing them in other headers accordingly.

I wouldn't move type definitions out of the headers, as like you say, that could break others' type references.

If there is any concern with header inclusion being cyclic, I think we should wrap the text of the headers in `ifndef` checks, e.g.:

```erlang
-ifndef(erlcloud_example_hrl).
-define(erlcloud_example_hrl).

# all header file definitions here

-endif.
```

That way, there's no issue with cross-references in headers that depend on other headers.

**Comment by paulo-ferraz-oliveira** [comment]
> in general, @paulo-ferraz-oliveira do i understand correct that your top level project requires dializer passing?

Yes. I try to have dialyzer pass with no warnings in all my projects, and their configuration is somewhat strict, in that sense. In the case of erlcloud, for example, the issue that I referred initially wouldn't prevent me from putting the code live, because the code compiles and it works (we test it before going live, of course). But if there's a warning, we try to handle it.

**Comment by paulo-ferraz-oliveira** [comment]
> That way, there's no issue with cross-references in headers that depend on other headers.

I'll make a PR for this.
